### ISSUE TITLE
Get rid of cast to int in IsFeatureOn

### ISSUE THREAD

**Issue by hotstone** [lloydstone/femah#26]
### Current situation

Currently a list of feature switches is declared using an enum: 

```
public enum MyFeatureSwitches {
   NewCheckout,
   GreenLayout
}
```

We then initialise Femah:

```
Femah.Configure()
    .FeatureSwitches(typeof(MyFeatureSwitches))
    .Initialise();
```

And then we can check a switch:

```
if ( Femah.IsFeatureOn( (int)MyFeatureSwitches.NewCheckout ) )
```

That cast to int is ugly but is currently required because at compile time the Femah class can't know what type the enum will be.  There are a couple of alternatives though.
### Alternative 1. Generic class in IsFeatureOn method

Declare the IsFeatureOn method with a generic type:

```
public static class Femah
{
    public bool IsFeatureOn<T>(T featureSwitch )
    {
       ...
```

And then to invoke:

```
if ( Femah.IsFeatureOn(MyFeatureSwitches.NewCheckout) ) 
```

The downside here is that you could equally invoke the method with any other parameter:

```
var obj = new Object();
if ( Femah.IsFeatureOn(obj) )
```
### Alternative 2. Generic static class

If we declare the static class using an anonymous type.

```
public static class Femah<T>
{
   bool IsFeatureOn(T featureSwitch) 
   {
      ...
```

Then the method of invoking IsFeatureOn would be:

```
if ( Femah<MyFeatureSwitches>.IsFeatureOn(MyFeatureSwitches.NewCheckout) )
```

The advantage here is that the parameter is now type checked, although of course the declaration of Femah is not:

```
if ( Femah<Biggles>.IsFeatureOn(myBigglesInstances) )
```

Another advantage is that you're not restricted to a single enum of feature switches.

Thoughts?

**Comment by lholman** [comment]
@hotstone I don't want to keep stretching out the MVP, but in your opinion would it be better to do this now, prior to the MVP release?  I think it would make documentation easier at least and would be nicer than changing later down the line?

**Comment by hotstone** [comment]
@lholman Agreed, this will be a massive and obvious change to make later.

I didn't make it clear above but the second approach suffers from the same disadvantage as the first alternative listed: You can still pass in any type you like each time you access Femah:

```
if ( Femah<MyFeatureSwitches>.IsFeatureOn(MyFeatureSwitches.NewCheckout) )
...

if ( Femah<SomeOtherType>.IsFeatureOn(new SomeOtherType())
...
```

which is nonsense.  It's also quite verbose with the enum needing to be listed twice.

**Comment by lholman** [comment]
@hotstone Yes, valid point. 

Is it that you're looking to make this nicer for the dev with Intellisense type safety, or is this a question of a trade-off with compilation or runtime exception handling?

All said, it's looking like Alternative 1 is the approach?

**Comment by hotstone** [comment]
@lholman Yep, thinking of the developer and intellisense etc.  Aiming for that so-called 'pit of success' where the system tends to direct the dev to the correct result.  It shouldn't affect runtime exception handling etc, although there would be unexpected results when the dev passes in an int from the wrong enum (it could correspond to a valid feature switch number by sheer coincidence).

All in all, I'd say 1 is the best approach.  It's nicer to read, and hasn't got any extra downsides that the current approach doesn't have.

**Comment by lholman** [comment]
@hotstone Ok, but it does tie you to a single enum of feature switches, right?

**Comment by hotstone** [comment]
@lholman Yep, it does.  Typecasting the Femah class frees you from that, but seems fairly verbose and implementing it will have a knock-on effect to the provider/featureswitch implementations (currently a single int uniquely identifies a featureswitch, but if you can have multiple enums then the unique identifier becomes type+int.)

**Comment by lholman** [comment]
@hotstone I don't really see a valid use case for multiple enums at present. Shall we go with Alternative 1, are you happy to pick it up?

**Comment by hotstone** [comment]
@lholman  Yep and yep.  I'll get cracking on it later today.
